main.tf

```
@@ -39,57 +39,86 @@ resource "aws_lambda_function" "lambda_asg" {
+resource "aws_iam_role" "asg_scheduler" {
+  count = var.asg_scheduler != null ? 1 : 0
+  name  = "${random_id.this[0].id}-asg-scheduler-role"
+  assume_role_policy = jsonencode({
+    Version = "2012-10-17"
+    Statement = [
+      {
+        Effect = "Allow"
+        Principal = {
+          Service = ["scheduler.amazonaws.com"]
+        }
+        Action = "sts:AssumeRole"
+      }
+    ]
   })
 }
 
-resource "aws_cloudwatch_event_rule" "asg_upscale_scheduler_event" {
-  count               = var.asg_scheduler != null ? 1 : 0
-  name                = "asg-scheduler-upscale-event-${random_id.this[0].id}"
-  description         = "Event rule for ASG upscale scheduler"
-  schedule_expression = var.asg_scheduler.upscale_cron_expression
+resource "aws_iam_role_policy_attachment" "asg_downscale_scheduler" {
+  count      = var.asg_scheduler != null ? 1 : 0
+  policy_arn = aws_iam_policy.asg_scheduler[0].arn
+  role       = aws_iam_role.asg_scheduler[0].name
 }
 
-resource "aws_lambda_permission" "allow_asg_upscale" {
-  count         = var.asg_scheduler != null ? 1 : 0
-  action        = "lambda:InvokeFunction"
-  function_name = aws_lambda_function.lambda_asg[0].function_name
-  principal     = "events.amazonaws.com"
-  source_arn    = aws_cloudwatch_event_rule.asg_upscale_scheduler_event[0].arn
+resource "aws_iam_policy" "asg_scheduler" {
+  count = var.asg_scheduler != null ? 1 : 0
+  name  = "${random_id.this[0].id}-asg-scheduler"
+  policy = jsonencode({
+    Version = "2012-10-17"
+    Statement = [
+      {
+        Effect = "Allow",
+        Action = [
+          "lambda:InvokeFunction"
+        ]
+        Resource = [aws_lambda_function.lambda_asg[0].arn]
+      }
+    ]
+  })
 }
 
-resource "aws_cloudwatch_event_target" "asg_upscale_scheduler_target" {
-  count     = var.asg_scheduler != null ? 1 : 0
-  rule      = aws_cloudwatch_event_rule.asg_upscale_scheduler_event[0].name
-  target_id = aws_lambda_function.lambda_asg[0].function_name
-  arn       = aws_lambda_function.lambda_asg[0].arn
-  input = jsonencode({
-    asg_name         = var.asg_scheduler.asg_name
-    desired_capacity = var.asg_scheduler.upscale_desired_capacity
-  })
+resource "aws_scheduler_schedule" "asg_downscale_scheduler" {
+  count = var.asg_scheduler != null ? 1 : 0
+  name  = "${random_id.this[0].id}-asg-downscale-schedule"
+
+  flexible_time_window {
+    mode = "OFF"
+  }
+
+  schedule_expression          = var.asg_scheduler.downscale_cron_expression
+  schedule_expression_timezone = "Europe/Zurich"
+
+  target {
+    arn      = aws_lambda_function.lambda_asg[0].arn
+    role_arn = aws_iam_role.asg_scheduler[0].arn
+    input = jsonencode({
+      asg_name         = var.asg_scheduler.asg_name
+      desired_capacity = var.asg_scheduler.downscale_desired_capacity
+    })
+  }
+}
+
+resource "aws_scheduler_schedule" "asg_upscale_scheduler" {
+  count = var.asg_scheduler != null ? 1 : 0
+  name  = "${random_id.this[0].id}-asg-upscale-schedule"
```

```suggestion
  name  = "asg-upscale-${random_id.this[0].id}"
```

Or because we ave multiple "aws_scheduler_schedule" resource in this module, then naming them like
name  = "asg-${random_id.this[0].id}-upscale"
name =  "asg-${random_id.this[0].id}-downscale"
...